go-controller/pkg/node/gateway_init.go

```diff
@@ -146,9 +148,45 @@ func (n *OvnNode) initGateway(subnets []*net.IPNet, nodeAnnotator kube.Annotator
 	// as that option does not add default SNAT rules on the GR and the gatewayReady function checks
 	// those default NAT rules are present
 	if !config.Gateway.DisableSNATMultipleGWs {
-		waiter.AddWait(gatewayReady, prFn)
+		waiter.AddWait(gatewayReady, gw.PostWait())
+	}
+
+	n.gateway = gw
+	_, err = n.watchFactory.AddServiceHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			svc := obj.(*kapi.Service)
+			n.gateway.AddService(svc)
+		},
+		UpdateFunc: func(old, new interface{}) {
+			oldSvc := old.(*kapi.Service)
+			newSvc := new.(*kapi.Service)
+			n.gateway.UpdateService(oldSvc, newSvc)
+		},
+		DeleteFunc: func(obj interface{}) {
+			svc := obj.(*kapi.Service)
+			n.gateway.DeleteService(svc)
+		},
+	}, n.gateway.SyncServices)
```

Comment:
This pattern is duplicated in far too many places in ovn-kubernetes. We should change the signatures of the factory functions so you can just do

    _, err = n.watchFactory.AddServiceHandler(n.gateway.AddService, n.gateway.UpdateService, n.gateway.DeleteService)

or

    _, err = n.watchFactory.AddServiceHandler(n.gateway)

(although that would be simpler here, I'm not sure it would be simpler globally...)

Likewise with `AddEndpointHandler`, etc. It could also allow the functions to return an error and then log it for them

Reply:
No, all I meant is that the `factory.go` functions should do all the type juggling for you. Instead of having

    func (wf *WatchFactory) AddServiceHandler(handlerFuncs cache.ResourceEventHandler, processExisting func([]interface{})) *Handler

we should have something like

```
type AddServiceFunc func(svc *kapi.Service)
type UpdateServiceFunc func(old, new *kapi.Service)
type DeleteServiceFunc func(svc *kapi.Service)
type SyncServicesFunc func(svcs []*kapi.Service)

func (wf *WatchFactory) AddServiceHandler(add AddServiceFunc, update UpdateServiceFunc, delete DeleteServiceFunc, sync SyncServicesFunc) *Handler {
	return wf.addHandler(serviceType, "", nil, cache.ResourceEventHandlerFuncs{
		AddFunc: func(obj interface{}) {
			svc := obj.(*kapi.Service)
			add(svc)
		},
		UpdateFunc: func(old, new interface{}) {
			oldSvc := old.(*kapi.Service)
			newSvc := new.(*kapi.Service)
			update(oldSvc, newSvc)
		},
		DeleteFunc: func(obj interface{}) {
			svc := obj.(*kapi.Service)
			delete(svc)
		},
	}, func(objs []interface{}) {
		// FIXME translate objs to svcs
		sync(svcs)
	})
}
```

and then the code above just becomes

    _, err = n.watchFactory.AddServiceHandler(n.gateway.AddService, n.gateway.UpdateService, n.gateway.DeleteService, n.gateway.SyncServices)

you can still have complicated add/update/delete/sync funcs, they just don't need to deal with the typecasting boilerplate